### PR TITLE
Fix broken highlight styles

### LINKED ISSUE #22348: Syntax Highlighting in `arc-green` Colors all Identifiers in Orange
### Description

From 1.17.1 to 1.17.2, syntax highlighting when viewing files in the `arc-green` theme changed in such a way that many parts of the code are assigned the same orange color. This is inconsistent with both the `gitea` theme and the way `arc-green` behaves when editing files.

For instance, in screenshot [1], regular identifiers use the regular text color, and special identifiers like function names and built-in types get a special color. In screenshot [2], all of these identifiers get the same orange color. In all of the code files I have looked at, the majority of the text was orange rather than the regular gray text color. This does not happen when editing files [3]. The light theme also uses regular text color for regular identifiers and special colors for special identifiers [4].

Syntax highlighting when viewing files in the `arc-green` theme seems to work the same way in `1.18.0` as it does in `1.17.2`

### Screenshots

[1] 1.17.1 - arc-green - view file
image
[2] 1.17.2 - arc-green - view file
image
[3] 1.17.2 - arc-green - edit file
image
[4] 1.17.2 - gitea - view file
image

### Gitea Version

1.17.2

### Can you reproduce the bug on the Gitea demo site?

Yes

### Operating System

_No response_

### Browser Version

Chromium Version 108.0.5359.124 (Official Build) Arch Linux (64-bit)

### PR DESCRIPTION
**Pull request by wxiaoguang**
Close #22348

This PR uses Chroma's style=github

image

image

### CODE REVIEW

**Comment by silverwind** [comment]
Need to modify chroma dark theme as well to really fix that issue.

**Comment by silverwind** [comment]
Can you make the background on dark the same color as the line numbers, like it is on light, probably by just not setting it?

**Comment by silverwind** [comment]
I think syntax theme should not set `.chroma` background color at all, that is the job of the gitea theme, not the syntax  theme.

**Comment by wxiaoguang** [comment]
> I think syntax theme should not set `.chroma` background at all, that is the job of the gitea theme, not the syntax theme.

https://github.com/go-gitea/gitea/pull/23148/commits/a4db9330ca8841fb1042aa451c99b87e80731ee3

**Comment by silverwind** [comment]
Also I would object to having `font-weight` and `font-style` being set at all in the theme. I perceive this as ugly, and Github also no longer does this so I would suggest dropping these declarations.

File: web_src/less/chroma/dark.less [hunk 1]
```diff
@@ -1,70 +1,75 @@
...
+/* stylelint-disable */
+
+/* PreWrapper */ .chroma { color: #c9d1d9; }
+/* Error */ .chroma .err { color: #f85149 }
+/* LineLink */ .chroma .lnlinks { outline: none; text-decoration: none; color: inherit }
+/* LineTableTD */ .chroma .lntd { vertical-align: top; padding: 0; margin: 0; border: 0; }
+/* LineTable */ .chroma .lntable { border-spacing: 0; padding: 0; margin: 0; border: 0; }
+/* LineHighlight */ .chroma .hl { color: #6e7681 }
+/* LineNumbersTable */ .chroma .lnt { white-space: pre; -webkit-user-select: none; user-select: none; margin-right: 0.4em; padding: 0 0.4em 0 0.4em;color: #64686c }
+/* LineNumbers */ .chroma .ln { white-space: pre; -webkit-user-select: none; user-select: none; margin-right: 0.4em; padding: 0 0.4em 0 0.4em;color: #6e7681 }
```
Comment (silverwind) [inline web_src/less/chroma/dark.less:12]: We do not use line numbers from chroma, remove these rules.

**Comment by wxiaoguang** [comment]
> Also I would object to having `font-weight` and `font-style` being set at all in the theme. I perceive this as ugly, and Github also no longer does this so I would suggest dropping these declarations.

This PR is only for fixing the broken chroma styles. Feel free to propose new PRs or replace this one.

File: web_src/less/chroma/dark.less (hunk 1 above)
Comment (wxiaoguang) [reply web_src/less/chroma/dark.less:12]: I do not think it's worth to do any more changes to the generated output, since it's not harmful.

**Comment by wxiaoguang** [comment]
Or please help to propose a complete document about how to follow/import Chroma's style correctly. Then future developers could know what to do.

**Comment by silverwind** [comment]
You are completely replacing the dark theme with another one and are adding these opinionated font styles from the chroma bundled themes which are not really matching GitHub. I do agree we can probably switch the dark color scheme to something more closely matching GitHub.

These Chroma-bundled themes are not suitable for us without more adjustments, like these font rules. I would also remove this comment on the generation, it's far from a simply copy-paste and needs more refinement.

**Comment by wxiaoguang** [comment]
https://github.com/go-gitea/gitea/pull/23148/commits/7dcee3240c2a167adee0911995234c6d52ff6261 could do the work, by overwriting `.chroma * { font styles }`

**Comment by wxiaoguang** [comment]
> These Chroma-bundled themes are not suitable for us without more adjustments, like these font rules. I would also remove this comment on the generation, it's far from a simply copy-paste and needs more refinement.

But #22348 has been open for a long time, maybe everyone is waiting for you to fine tune the theme since the last modification was also contributed by you. I guess nobody else knows how to get these styles.

**Comment by wxiaoguang** [comment]
> [7dcee32](https://github.com/go-gitea/gitea/commit/7dcee3240c2a167adee0911995234c6d52ff6261) could do the work, by overwriting `.chroma * { font styles }`

Actually, I think the `bold` is just fine and done by purpose, some keywords are just "black + bold".  Forcing `font-weight: normal` also makes the UI looks strange.

I would suggest to:

1. Follow a major Chroma style, do not change too much
3. Write a clear document about how to maintain these styles

**Comment by wxiaoguang** [comment]
@silverwind 

Could you take a look at https://swapoff.org/chroma/playground/ and choose 2 best styles (light/dark) and just use them? I do not think it's worth to fine tune everything on our side. These styles are contributed by Chroma users and have been tested and used heavily, it could be better than our home-made styles. And using these style will reduce maintaining work, there will be no bug in the future even if Chroma changes its output again.

**Comment by silverwind** [comment]
I do not like the chroma styles for various reasons, one being the bold/italic. I think we can generate our own dark/light styles based on GitHub's colors available via https://github.com/primer/primitives. All that's left to do is convert prettylight primitives to chroma.

```dark
    --color-prettylights-syntax-comment: #8b949e;
    --color-prettylights-syntax-constant: #79c0ff;
    --color-prettylights-syntax-entity: #d2a8ff;
    --color-prettylights-syntax-storage-modifier-import: #c9d1d9;
    --color-prettylights-syntax-entity-tag: #7ee787;
    --color-prettylights-syntax-keyword: #ff7b72;
    --color-prettylights-syntax-string: #a5d6ff;
    --color-prettylights-syntax-variable: #ffa657;
    --color-prettylights-syntax-brackethighlighter-unmatched: #f85149;
    --color-prettylights-syntax-invalid-illegal-text: #f0f6fc;
    --color-prettylights-syntax-invalid-illegal-bg: #8e1519;
    --color-prettylights-syntax-carriage-return-text: #f0f6fc;
    --color-prettylights-syntax-carriage-return-bg: #b62324;
    --color-prettylights-syntax-string-regexp: #7ee787;
    --color-prettylights-syntax-markup-list: #f2cc60;
    --color-prettylights-syntax-markup-heading: #1f6feb;
    --color-prettylights-syntax-markup-italic: #c9d1d9;
    --color-prettylights-syntax-markup-bold: #c9d1d9;
    --color-prettylights-syntax-markup-deleted-text: #ffdcd7;
    --color-prettylights-syntax-markup-deleted-bg: #67060c;
    --color-prettylights-syntax-markup-inserted-text: #aff5b4;
    --color-prettylights-syntax-markup-inserted-bg: #033a16;
    --color-prettylights-syntax-markup-changed-text: #ffdfb6;
    --color-prettylights-syntax-markup-changed-bg: #5a1e02;
    --color-prettylights-syntax-markup-ignored-text: #c9d1d9;
    --color-prettylights-syntax-markup-ignored-bg: #1158c7;
    --color-prettylights-syntax-meta-diff-range: #d2a8ff;
    --color-prettylights-syntax-brackethighlighter-angle: #8b949e;
    --color-prettylights-syntax-sublimelinter-gutter-mark: #484f58;
    --color-prettylights-syntax-constant-other-reference-link: #a5d6ff;
```

light:
```
    --color-prettylights-syntax-comment: #6e7781;
    --color-prettylights-syntax-constant: #0550ae;
    --color-prettylights-syntax-entity: #8250df;
    --color-prettylights-syntax-storage-modifier-import: #24292f;
    --color-prettylights-syntax-entity-tag: #116329;
    --color-prettylights-syntax-keyword: #cf222e;
    --color-prettylights-syntax-string: #0a3069;
    --color-prettylights-syntax-variable: #953800;
    --color-prettylights-syntax-brackethighlighter-unmatched: #82071e;
    --color-prettylights-syntax-invalid-illegal-text: #f6f8fa;
    --color-prettylights-syntax-invalid-illegal-bg: #82071e;
    --color-prettylights-syntax-carriage-return-text: #f6f8fa;
    --color-prettylights-syntax-carriage-return-bg: #cf222e;
    --color-prettylights-syntax-string-regexp: #116329;
    --color-prettylights-syntax-markup-list: #3b2300;
    --color-prettylights-syntax-markup-heading: #0550ae;
    --color-prettylights-syntax-markup-italic: #24292f;
    --color-prettylights-syntax-markup-bold: #24292f;
    --color-prettylights-syntax-markup-deleted-text: #82071e;
    --color-prettylights-syntax-markup-deleted-bg: #ffebe9;
    --color-prettylights-syntax-markup-inserted-text: #116329;
    --color-prettylights-syntax-markup-inserted-bg: #dafbe1;
    --color-prettylights-syntax-markup-changed-text: #953800;
    --color-prettylights-syntax-markup-changed-bg: #ffd8b5;
    --color-prettylights-syntax-markup-ignored-text: #eaeef2;
    --color-prettylights-syntax-markup-ignored-bg: #0550ae;
    --color-prettylights-syntax-meta-diff-range: #8250df;
    --color-prettylights-syntax-brackethighlighter-angle: #57606a;
    --color-prettylights-syntax-sublimelinter-gutter-mark: #8c959f;
    --color-prettylights-syntax-constant-other-reference-link: #0a3069;
```

**Comment by wxiaoguang** [comment]
> I do not like the chroma styles for various reasons, one being the bold/italic. 

There are many Chroma styles without bold/italic, and the bold/italic styles could be fixed manually by adding a few new lines, without changing the original output too much.


> All that's left to do is convert the prettylights primitives to chroma.

It's not a 1v1 mapping to Chroma. If Chroma changes its output in the future (like this time), people have to  learn to do this again. I don't like to maintain the color mapping between Chroma and an external color system. But if most people like that mapping and have implemented it and have a document, then I would have no objection.

**Comment by mrsdizzie** [comment]
I only worked on the light theme aspect and can't speak to anything related to the dark theme or the linked issue --- which seems like a real issue to be fixed. 

For what it is worth these Chroma styles were considered and intentionally not used as-is when adding Chroma syntax highlighting because I (and others) found them to be both "not that nice looking" and to not reflect what Github looked like at the time. It doesn't look like they have changed at all since then. I know looks are an opinion, but just as a background that this option was considered and rejected previously because it didn't seem good enough for Gitea users.

The light Chroma github style was used as an initial baseline, some parts were updated base on https://github.com/primer/github-syntax-theme-generator and other things were probably added manually to match what Github was actually doing based on looking at lots of examples. 

"Fix broken highlight styles" seems like a misleading PR title here as the linked issue is related to an aspect of the dark theme (?) but this changes the light theme as well and isn't just a fix to current styles. The title and description should be changed to reflect what is really happening including before and after images which I think would show that this update feels like a visual regression at least in the light theme.

**Comment by wxiaoguang** [comment]
@mrsdizzie I basically agree with you. My opinions are:

* I do not think we need to strictly follow GitHub, although it's also good to follow if it doesn't make maintainability  more difficult.
* If there is a "more or less" nice theme from Chroma, we could just use it, and we just need to do some fine tunes.
* If we want to follow GitHub strictly and replace all colors, there should be a very clear document about what to do. Otherwise if the styles get broken again, nobody knows what to do next time.
* Why that bug is open for long time? Maybe nobody knows how to do.
* If I understand correctly, the "light theme" might be also broken somewhat. According to my last review, I can see  that the styles doesn't match the Chroma generated  ones.  I can confirm  that some of the missing styles are not used, while I  can not understand whether some others are used or not. That's why I planned to re-generated and re-sync the styles from Chroma outputs.

**Comment by wxiaoguang** [comment]
The alternative PR for fully customized styles :  Fix broken Chroma CSS styles #23174

**Pull request closed without merging**